### ISSUE TITLE
Add an isoslice function

### ISSUE THREAD

**Issue by stefraynaud** [VACUMM/xoa#49]
* Must work with numba/apply_ufunc
* Must be put in interp

**Comment by Matt29** [comment]
Hi stéphane,
As you suggested, i added your function _isoslice_ in **interp.py** and an _xisolice_ function in **regrid.py** compatible with _apply_ufunc_ so it returns an xarray object.
I'm not sure the xioslice is robust enough. Maybe some tests are needed before the call of _apply_ufunc_ ?
What do you think about it ?

```

@numba.guvectorize([(numba.float64[:], numba.float64[:], numba.float64[:], numba.float64[:])], "(nz),(nz),()->()")
def isoslice(var, values, isoval, isovar):
    """ Extract data from var where values==isoval

    Parameters
    -----------
    var: array_like
          array from which the data are extracted
    values: array_like
          array on which a research of isoval is made
    isoval: Float
          value of interest on which we perform research in values array 
    
    Return
    ------
    isovar : array_like 
           Sliced array based on var where values==isoval

    Example
    -------
   
    Let's define depth and temperature variables both in 3 dimensions (i,j,k)
    where i and j are horizontal dimension and k the vertical one.

    dep_at_t20 = isoslice(dep,temp,20)   # depth at temperature=20°C
    temp_at_z15 = isoslice(temp,dep,-15) # temperature at depth=-15m     
   

    """
    nz = var.shape[-1]
    isovar[0] = var[-1]

    # From the top
    for i in numba.prange(nz - 1, 0, -1):
        if (values[i] >= isoval[0] and values[i - 1] <= isoval[0]) or (
            values[i] <= isoval[0] and values[i - 1] >= isoval[0]
        ):
            if values[i - 1] == values[i]:
                isovar[0] = values[i]
            else:
                isovar[0] = var[i - 1] + (isoval[0] - values[i - 1]) * (
                    var[i] - var[i - 1]
                ) / (values[i] - values[i - 1])
```
```
def xisoslice(da,values,isoval,dim,**kwargs):
    """ Extract data from var where values==isoval

    Parameters
    -----------
    da: xarray.DataArray
          array from which the data are extracted
    values: array_like
          array on which a research of isoval is made
    isoval: Float
          value of interest on which we perform research in values array 
    dim   : str 
          dimension shared by da and values on which the slice is made
    Return
    ------
    isovar : array_like 
            Sliced array based on data where values==isoval

    Example
    -------
   
    Let's define depth and temperature variables both in 3 dimensions (i,j,k)
    where i and j are horizontal dimension and k the vertical one.

    dep_at_t20 = isoslice(dep,temp,20)   # depth at temperature=20°C
    temp_at_z15 = isoslice(temp,dep,-15) # temperature at depth=-15m     
   

    """

    assert dim in da.dims
    assert dim in values.dims

    da_out = xr.apply_ufunc(
        interp.isoslice,
        da,
        values,
        isoval,
        join="override",
        input_core_dims=[[dim],[dim],[]],
        exclude_dims={dim},
        dask='parallelized',
        dask_gufunc_kwargs={"output_sizes": da.sizes},
        **kwargs,
    )
    da_out.attrs.update(da.attrs)
    da_out.encoding.update(da.encoding)
    return da_out


```

**Comment by stefraynaud** [comment]
Excellent !

`interp.py` only contains low level routines with no conflict in names with other modules, so we should probably name the xarray-compatible routine `isoslice` also (or name the numba based routine `core_isoslice` for example).

The `dim` keyword can probably be set to `None` by default and guessed by `xoa.coords.get_zdim` in this case.

Try the routine on a sample a dataset that is included with the sources, like the croco one. It can be an extension to the existing doc example about croco.

Don't hesitate to apply black to your code.

**Comment by Matt29** [comment]
Thanks for the tips , i didn't know black !
I added an example of isoslice in `plot_croco_section.py`  and created a branch **dev_isoslice**
I built the doc in local and it's okay .

I think i'm going to make a PR now

**Comment by stefraynaud** [comment]
Closed by #63
